cwltool/provenance.py

```diff
@@ -0,0 +1,310 @@
+            else:
+                relativised_input_objecttemp[key]=value
+        relativised_input_object.update({k: v for k, v in relativised_input_objecttemp.items() if v})
+        return relativised_input_object
+
+    def _relativise_files(self, structure, kwargs):
+        '''
+        save any file objects into RO and update the local paths
+        '''
+        provenance_input_params={}
+        # Base case - we found a File we need to update
+        _logger.debug(u"[provenance] Relativising: %s", structure)
+        if isinstance(structure, dict):
+            if structure.get("class") == "File":
+                #standardised fs access object creation
+                fsaccess = kwargs["make_fs_access"]("")
+                # TODO: Replace location/path with new add_data_file() paths
+                with fsaccess.open(structure["location"], "rb") as f:
+                    relative_path=self.add_data_file(f)
+                    structure["location"]= "../"+relative_path
+
+            for o in structure.values():
+                self._relativise_files(o, kwargs)
+            return
+
+        if isinstance(structure, str) or isinstance(structure, unicode):
+            # Just a string value, no need to iterate further
+            return
+        try:
+            for o in iter(structure):
+                # Recurse and rewrite any nested File objects
+                self._relativise_files(o, kwargs)
+        except TypeError:
+            pass
+#**************************************
+    #copy output files to the RO
+    def add_output(self, workflow_output=None, saveTo=None):
+        if isinstance(workflow_output, dict):
+            #Iterate over the output object, collect and relativise the output file paths
+            for item in workflow_output:
+                if workflow_output[item]["class"] == "File":
+                    outputfile_path= os.path.join(self.folder, OUTPUT, workflow_output[item]["checksum"][5:7])
+                    path = os.path.join(outputfile_path, workflow_output[item]["checksum"][5:])
+                    if not os.path.isdir(path):
+                        os.makedirs(path)
+                    _logger.info(u"[provenance] Moving output files to RO")
+                    shutil.copy(workflow_output[item]["location"][7:], path)
+
+#**************************************
+    def add_provProfile(self, document):
+        '''
+        Transfer the provenance related files to RO
+        '''
+        original_path=os.path.join(self.folder, PROVENANCE)
+        provPath=original_path+"/ProvenanceProfile.json"
+        provDot=original_path+'/ProvenanceDotGraph.dot'
+        document.serialize(provPath, indent=2)
+        provgraph=graph.prov_to_graph(document)
+        pos = nx.nx_agraph.graphviz_layout(provgraph)
+        nx.draw(provgraph, pos=pos)
+        write_dot(provgraph, provDot)
+        check_call(['dot','-Tpng',provDot,'-o',original_path+'/ProvenanceVisualGraph.png'])
```

Comment:
Mind if I make the graphviz dependency optional? I can make a PR to this PR and you can review it

Reply:
Yes, that is fine I think. We can add Prov Viewer (https://github.com/gems-uff/prov-viewer) later.